tests/data_templates/test_upload_fields.py

```diff
@@ -0,0 +1,144 @@
+from io import StringIO
+import os
+import json
+import pytest
+from graphql import GraphQLError
+from marshmallow import ValidationError
+
+from creator.data_templates.models import TemplateVersion
+
+FIELDS = {"fields": [
+    {
+        "key": "person.id",
+        "label": "Person ID",
+        "description": "Person identifiers",
+        "required": True,
+        "data_type": "string",
+        "instructions": f"Populate person id properly",
+    }
+]}
+
+UPLOAD_FIELDS_MUTATION = """
+    mutation ($file: Upload!) {
+      uploadFieldDefinitions(file:$file) {
+        success
+        fileName
+        fileSize
+        fieldDefinitions
+      }
+    }
+"""
+
+
+@pytest.mark.parametrize(
+    "permissions,allowed",
+    [
+        (["change_datatemplate"], True),
+        ([], False),
+    ],
+)
+def test_upload_field_definitions(
+    db, tmpdir, permission_client, permissions, allowed
+):
+    """
+    Test the upload field definition mutation
+
+    Users without the change_datatemplate permission should not be allowed
+    to upload field definitions
+    """
+    user, client = permission_client(permissions)
+
+    # Make temp file and upload it
+    fields_file = os.path.join(tmpdir.mkdir("test"), "fields.json")
+    with open(fields_file, "w+") as json_file:
+        json.dump(FIELDS, json_file)
+        json_file.seek(0)
+        data = {
+            "operations": json.dumps(
+                {
+                    "query": UPLOAD_FIELDS_MUTATION.strip(),
+                    "variables":  {"file": None}
+                }
+            ),
+            "file": json_file,
+            "map": json.dumps({"file": ["variables.file"]}),
```

Comment:
what is this?

Reply:
Ugh yea its weird. Its part of the graphql multipart request spec: https://github.com/jaydenseric/graphql-multipart-request-spec#single-file. 

I think the `map` value is basically saying look at the `file` variable in the `variables` part of the request content in order to read the file content from the file stream